arangod/VocBase/ComputedValues.h

```diff
@@ -0,0 +1,210 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Aql/AqlFunctionsInternalCache.h"
+#include "Aql/AqlValue.h"
+#include "Aql/ExpressionContext.h"
+#include "Basics/ErrorCode.h"
+#include "Basics/Result.h"
+#include "Containers/FlatHashMap.h"
+#include "Containers/FlatHashSet.h"
+
+#include <cstdint>
+#include <string>
+#include <string_view>
+#include <vector>
+
+struct TRI_vocbase_t;
+
+namespace arangodb {
+class LogicalCollection;
+struct ValidatorBase;
+
+namespace aql {
+struct AstNode;
+class Expression;
+class ExpressionContext;
+class QueryContext;
+struct Variable;
+}  // namespace aql
+
+namespace transaction {
+class Methods;
+}
+
+namespace velocypack {
+class Builder;
+class Slice;
+}  // namespace velocypack
+
+enum class ComputeValuesOn : uint8_t {
+  kNever = 0,
+  kInsert = 1,
+  kUpdate = 2,
+  kReplace = 4,
+};
+
+// expression context used for calculating computed values inside
+class ComputedValuesExpressionContext final : public aql::ExpressionContext {
+ public:
+  explicit ComputedValuesExpressionContext(transaction::Methods& trx,
+                                           LogicalCollection& collection);
+
+  void registerWarning(ErrorCode errorCode, char const* msg) override;
```

Comment:
IMO this is a valid point of view.
But my point of view is this:
As the class is currently `final` there is no immediate benefit in marking all methods as `final` again. There would only be benefit if someone removes the `final` specifier from the class, which right now seems unlikely. And when the `final` specifier gets removed from the class, IMO it is even better that not every method is marked as final individually, because a `final` specifier gets removed in order to override things. 

I think both view points are reasonable. So what shall we do?

Reply:
I think for this PR do nothing (for faster merge and lesser code changes).

But in general maybe we can make poll, choose one option, put it in codestyle, and try to follow it.

Btw as another argument to suggested option:
it more consistent, final methods always marked as final, don't need to check it in class, more easy to find, etc

Reply:
Running a poll is a good idea. I can live with both outcomes, but I also think we should agree on one variant for consistency.